src/Mod/Import/gltfGlbExt.py

```diff
@@ -0,0 +1,371 @@
+        json_str += b" " * json_padding
+
+        header_size = 12  # GLB header
+        json_chunk_header_size = 8  # JSON chunk header
+
+        original_json_length = struct.unpack("<I", original_data[12:16])[0]
+
+        binary_start = header_size + json_chunk_header_size + original_json_length
+
+        binary_data = original_data[binary_start:]
+
+        new_total_length = header_size + json_chunk_header_size + len(json_str) + len(binary_data)
+
+        with open(filename, "wb") as f:
+            f.write(b"glTF")
+            f.write(struct.pack("<I", 2))
+            f.write(struct.pack("<I", new_total_length))
+
+            f.write(struct.pack("<I", len(json_str)))
+            f.write(b"JSON")
+
+            f.write(json_str)
+
+            if binary_data:
+                f.write(binary_data)
+
+    except (IOError, OSError, struct.error, json.JSONEncodeError, UnicodeEncodeError):
+        pass
+
+
+def get_all_objects_recursive(objects):
+    """
+    Recursively traverse all objects including sub-groups to find all objects.
+
+    Args:
+        objects: List of FreeCAD objects to traverse
+
+    Returns:
+        List of all objects found (including nested ones)
+    """
+    all_objects = []
+
+    def traverse_object(obj):
+        all_objects.append(obj)
+
+        if hasattr(obj, "Group") and obj.Group:
+            for sub_obj in obj.Group:
+                traverse_object(sub_obj)
+
+        if hasattr(obj, "OutList") and obj.OutList:
+            for sub_obj in obj.OutList:
+                traverse_object(sub_obj)
+
+        if hasattr(obj, "LinkedObject") and obj.LinkedObject:
+            traverse_object(obj.LinkedObject)
+
+        if hasattr(obj, "Objects") and obj.Objects:
+            for sub_obj in obj.Objects:
+                traverse_object(sub_obj)
+
+    for obj in objects:
+        traverse_object(obj)
+
+    # Remove duplicates while preserving order
+    seen = set()
+    unique_objects = []
+    for obj in all_objects:
+        if id(obj) not in seen:
+            seen.add(id(obj))
+            unique_objects.append(obj)
+
+    return unique_objects
+
+
+def export(objects, filename):
+    """
+    GLTF/GLB export extension function with material name preservation.
+    """
+    processed_objects = preprocess_objects(objects)
+
+    try:
+        ImportGui.export(processed_objects, filename)
+        postprocess_export(filename)
+    except Exception as e:
+        raise
+
+
+def preprocess_objects(objects):
+    """
+    Preprocess objects and collect material names for GLTF export.
+    """
+    global found_material_names
+    found_material_names = []
+
+    all_objects = get_all_objects_recursive(objects)
+    collect_all_material_names(all_objects)
+
+    # Collect VarSet data for GLTF injection
+    global collected_varset_data
+    collected_varset_data = {}
+
+    varset_paths = find_varsets_with_path(objects)
+    for varset_obj, full_path in varset_paths:
```

Comment:
Consider breaking some of these nested loops out into sub-functions for readability.